### PR TITLE
Add MemoryTracker for the background tasks [Resubmit]

### LINKED ISSUE #48774: ThreadSanitizer: use of an invalid mutex (e.g. uninitialized or destroyed)
https://s3.amazonaws.com/clickhouse-test-reports/45819/d17abc9e89990dc36594a178d4cff6fbff2bf414/stress_test__tsan_.html

```
WARNING: ThreadSanitizer: use of an invalid mutex (e.g. uninitialized or destroyed) (pid=38762)
    #0 pthread_mutex_lock <null> (clickhouse+0xce2be9a) (BuildId: eeeb4b8c47df3a26a45b773beefba242cb57015a)
    #1 Poco::MutexImpl::lockImpl() build_docker/./base/poco/Foundation/include/Poco/Mutex_POSIX.h:60:9 (clickhouse+0x1cd08050) (BuildId: eeeb4b8c47df3a26a45b773beefba242cb57015a)
    #2 Poco::Mutex::lock() build_docker/./base/poco/Foundation/include/Poco/Mutex.h:198:5 (clickhouse+0x1cd08050)
    #3 Poco::ScopedLock<Poco::Mutex>::ScopedLock(Poco::Mutex&) build_docker/./base/poco/Foundation/include/Poco/ScopedLock.h:37:61 (clickhouse+0x1cd08050)
    #4 Poco::Logger::get(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) build_docker/./base/poco/Foundation/src/Logger.cpp:285:20 (clickhouse+0x23fa129f) (BuildId: eeeb4b8c47df3a26a45b773beefba242cb57015a)
    #5 MemoryTracker::logPeakMemoryUsage() build_docker/./src/Common/MemoryTracker.cpp:126:5 (clickhouse+0x14f5d59e) (BuildId: eeeb4b8c47df3a26a45b773beefba242cb57015a)
    #6 MemoryTracker::~MemoryTracker() build_docker/./src/Common/MemoryTracker.cpp:113:13 (clickhouse+0x14f5d4f2) (BuildId: eeeb4b8c47df3a26a45b773beefba242cb57015a)
    #7 cxa_at_exit_callback_installed_at(void*) crtstuff.c (clickhouse+0xce4b1ca) (BuildId: eeeb4b8c47df3a26a45b773beefba242cb57015a)
    #8 __cxx_global_var_init.3 build_docker/./src/Common/MemoryTracker.cpp (clickhouse+0x14f61a3f) (BuildId: eeeb4b8c47df3a26a45b773beefba242cb57015a)
    #9 _GLOBAL__sub_I_MemoryTracker.cpp build_docker/./src/Common/MemoryTracker.cpp (clickhouse+0x14f61a3f)

  Location is global 'Poco::Logger::_mapMtx' of size 40 at 0x00002949c9d0 (clickhouse+0x2949c9d0)

  Mutex M0 (0x00002949c9d0) created at:
    [failed to restore the stack]

SUMMARY: ThreadSanitizer: use of an invalid mutex (e.g. uninitialized or destroyed) (/usr/bin/clickhouse+0xce2be9a) (BuildId: eeeb4b8c47df3a26a45b773beefba242cb57015a) in pthread_mutex_lock
```

### PR DESCRIPTION
**Pull request by novikd**
### Changelog category (leave one):

- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add MemoryTracker for the background tasks (merges and mutation). Introduces `merges_mutations_memory_usage_soft_limit` and `merges_mutations_memory_usage_to_ram_ratio` settings that represent the soft memory limit for merges and mutations. If this limit is reached ClickHouse won't schedule new merge or mutation tasks. Also `MergesMutationsMemoryTracking` metric is introduced to allow observing current memory usage of background tasks.
Resubmit #46089. Closes #48774.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Review by CheSema — APPROVED** [review]
Diff looks OK.
Some tests are failed.

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit 93e8ecf4ffe6b5766792c2316fc024b4f9c05f84 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/48787/93e8ecf4ffe6b5766792c2316fc024b4f9c05f84/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

File: src/Common/MemoryTracker.cpp [hunk 1]
```diff
@@ -96,12 +96,17 @@ using namespace std::chrono_literals;
 static constexpr size_t log_peak_memory_usage_every = 1ULL << 30;
 
 MemoryTracker total_memory_tracker(nullptr, VariableContext::Global);
+MemoryTracker background_memory_tracker(&total_memory_tracker, VariableContext::User, false);
```
[the remark this answers is not in the document]
Comment (novikd) [reply src/Common/MemoryTracker.cpp:99]: We don't have an order when global objects are destroyed.

[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/MemoryTracker.cpp:99]: background_memory_tracker do some staff in d-tor. That might be the problem I think. It calls some object which could be destroyed already.

**Pull request merged**